File: 0075-sort-colors/0075-sort-colors.java
class Solution {
    public void sortColors(int[] nums) {
        int low = 0; // Pointer for next position of 0
        int mid = 0; // Pointer for current element
        int high = nums.length - 1; // Pointer for next position of 2

        while (mid <= high) {
            if (nums[mid] == 0) {
                swap(nums, low, mid);
                low++;
                mid++;
            } else if (nums[mid] == 1) {
                mid++;
            } else { // nums[mid] == 2
                swap(nums, mid, high);
                high--;
            }
        }
    }

    private void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
}

File: 0075-sort-colors/0075-sort-colors.py
class Solution:
    def sortColors(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        low = 0            # Pointer for the next position of 0
        mid = 0            # Pointer for the current element
        high = len(nums) - 1  # Pointer for the next position of 2

        while mid <= high:
            if nums[mid] == 0:
                # Swap nums[low] and nums[mid], then increment both
                nums[low], nums[mid] = nums[mid], nums[low]
                low += 1
                mid += 1
            elif nums[mid] == 1:
                # If it's 1, just move to the next element
                mid += 1
            else:  # nums[mid] == 2
                # Swap nums[mid] and nums[high], then decrement high
                nums[mid], nums[high] = nums[high], nums[mid]
                high -= 1

